python/cudf/cudf/core/udf/row_function.py

```diff
@@ -0,0 +1,171 @@
+import math
+
+import numpy as np
+from numba import cuda
+from numba.np import numpy_support
+from numba.types import Record
+
+from cudf.core.udf.api import Masked, pack_return
+from cudf.core.udf.templates import (
+    masked_input_initializer_template,
+    row_initializer_template,
+    row_kernel_template,
+    unmasked_input_initializer_template,
+)
+from cudf.core.udf.typing import MaskedType
+from cudf.core.udf.utils import (
+    _all_dtypes_from_frame,
+    _construct_signature,
+    _get_kernel,
+    _get_udf_return_type,
+    _mask_get,
+    _supported_cols_from_frame,
+    _supported_dtypes_from_frame,
+)
+
+
+def _get_frame_row_type(dtype):
+    """
+    Get the numba `Record` type corresponding to a frame.
+    Models each column and its mask as a MaskedType and
+    models the row as a dictionary like data structure
+    containing these MaskedTypes.
+
+    Large parts of this function are copied with comments
+    from the Numba internals and slightly modified to
```

Comment:
Is there somewhere we can link to for these numba internals?

Reply:
Added a note about this below